sdk/monitor/azure-monitor-ingestion/src/main/java/com/azure/monitor/ingestion/implementation/Batcher.java

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+// Code generated by Microsoft (R) AutoRest Code Generator.
+
+package com.azure.monitor.ingestion.implementation;
+
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.core.util.serializer.JsonSerializer;
+import com.azure.core.util.serializer.JsonSerializerProviders;
+import com.azure.core.util.serializer.ObjectSerializer;
+import com.azure.monitor.ingestion.models.LogsUploadOptions;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+import reactor.core.publisher.Flux;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static com.azure.monitor.ingestion.implementation.Utils.MAX_REQUEST_PAYLOAD_SIZE;
+import static com.azure.monitor.ingestion.implementation.Utils.getConcurrency;
+import static com.azure.monitor.ingestion.implementation.Utils.gzipRequest;
+
+/**
+ *  Provides iterator and streams for batches over log objects.
+ */
+public class Batcher implements Iterator<LogsIngestionRequest> {
+    private static final ClientLogger LOGGER = new ClientLogger(Utils.class);
+    private static final JsonSerializer DEFAULT_SERIALIZER = JsonSerializerProviders.createInstance(true);
+    private final ObjectSerializer serializer;
+    private final int concurrency;
+    private final Iterator<Object> iterator;
+    private long currentBatchSize;
+    private List<String> serializedLogs;
+    private List<Object> originalLogsRequest;
+
+    public Batcher(LogsUploadOptions options, Iterable<Object> iterable) {
+        this.serializer = getSerializer(options);
+        this.concurrency = getConcurrency(options);
+        this.serializedLogs = new ArrayList<>();
+        this.originalLogsRequest = new ArrayList<>();
+        this.iterator = iterable.iterator();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return iterator.hasNext() || currentBatchSize > 0;
+    }
+
+    /**
+     * Collects next batch and serializes it into {@link LogsIngestionRequest}. This method is not thread-safe!
```

Comment:
Good to note that this is not thread-safe! We should do the same for `hasNext()` too.